src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1964,19 +1966,30 @@ def _process_decln(self, scope, symbol_table, decl, visibility_map=None):
             tag = None
             try:
                 sym = symbol_table.lookup(sym_name, scope_limit=scope)
-                if sym is symbol_table.lookup_with_tag("own_routine_symbol"):
-                    # In case it is its own function routine symbol, Fortran
-                    # will declare it inside the function as a DataSymbol.
-                    # Remove the RoutineSymbol in order to free the exact name
-                    # for the DataSymbol.
-                    symbol_table.remove(sym)
-                    # And trigger the exception path but keeping the same tag
-                    tag = "own_routine_symbol"
-                    raise KeyError
-                if not sym.is_unresolved:
-                    raise SymbolError(
-                        f"Symbol '{sym_name}' already present in SymbolTable "
-                        f"with a defined interface ({sym.interface}).")
+                if type(sym) == Symbol:
+                    # This was a generic symbol. We now know what it is
+                    sym.specialise(DataSymbol, datatype=datatype,
+                                   visibility=visibility,
+                                   is_constant=has_constant_value,
+                                   initial_value=init_expr)
```

Comment:
and sym.interface=interface? so it won't be unresolved anymore. I don't know if it is accepted as a specialise parameter.

Reply:
It is accepted, but was not properly supported in `_process_decln`. This is hopefully fixed now.